# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: docshell/base/BrowsingContext.h
# Commit: cefac6461948
# Full Hash: cefac6461948d6b8ef6cd6b1f563a2e487e0bbac
# Author: Kris Maglione <maglione.k@gmail.com>
# Date: 2020-02-07 03:55:51
# Regressor Bug: 1582832
# File Overlap Count: 1
# Description:
#   Bug 1582832: Part 2 - Don't attach a BrowsingContext before a FrameLoader is initialized. r=nika
#   
#   There are cases where the initial BrowsingContext that we create for a
#   FrameLoader will never have a DocShell created for it. In the particular case
#   when the frame it belongs to is part of an inactive document, trying to
# ==============================================================================

diff -r 6fd57b83bb52 -r cefac6461948 docshell/base/BrowsingContext.h
--- a/docshell/base/BrowsingContext.h	Thu Feb 06 19:07:56 2020 +0000
+++ b/docshell/base/BrowsingContext.h	Thu Feb 06 19:08:04 2020 +0000
@@ -155,6 +155,17 @@
                                                   const nsAString& aName,
                                                   Type aType);
 
+  // Same as the above, but does not immediately attach the browsing context.
+  // `EnsureAttached()` must be called before the BrowsingContext is used for a
+  // DocShell, BrowserParent, or BrowserBridgeChild.
+  static already_AddRefed<BrowsingContext> CreateDetached(
+      BrowsingContext* aParent, BrowsingContext* aOpener,
+      const nsAString& aName, Type aType);
+
+  void EnsureAttached();
+
+  bool EverAttached() const { return mEverAttached; }
+
   // Cast this object to a canonical browsing context, and return it.
   CanonicalBrowsingContext* Canonical();
 
@@ -634,6 +645,9 @@
   JS::Heap<JSObject*> mWindowProxy;
   LocationProxy mLocation;
 
+  // True if Attach() has been called on this BrowsingContext already.
+  bool mEverAttached : 1;
+
   // Is the most recent Document in this BrowsingContext loaded within this
   // process? This may be true with a null mDocShell after the Window has been
   // closed.